src: add .code and SSL specific error properties

sam-github opened the pull request:
SSL errors have a long structured message, but lacked the standard .code
property which can be used for stable comparisons. Add a `code`
property, as well as the 3 string components of an SSL error: `reason`,
`library`, and `function`.

This is WIP, TBD is move the property strings to the isolate, and add docs and more tests.

I've seen a reasonable amount of code that is doing string matching on the error string, looking for the reason part, I think exposing the properties makes more sense. Also, I think the intention is for all Errors to expose `.code` properties for stable programmatic inspection.

I've gone some way to exposing these properties for node_crypto.cc, too, but I'm not done. Once this PR lands I'll keep working on it.

/to @nodejs/crypto 


<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [ ] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

jasnell commented on src/tls_wrap.cc:
```diff
@@ -356,6 +357,40 @@ Local<Value> TLSWrap::GetSSLError(int status, int* err, std::string* msg) {
+        const char* fs = ERR_func_error_string(e);
+        const char* rs = ERR_reason_error_string(e);
+
+        if (ls != nullptr) obj->Set(env()->isolate()->GetCurrentContext(),
```
nit: Just for DRY... perhaps ...

```c
Local<Context> context = env()->isolate()->GetCurrentContext();
// ...
if (ls != nullptr) obj->Set(context, ...)
```

jasnell commented on src/tls_wrap.cc:
```diff
@@ -356,6 +357,40 @@ Local<Value> TLSWrap::GetSSLError(int status, int* err, std::string* msg) {
+                 OneByteString(env()->isolate(), "reason"),
+                 OneByteString(env()->isolate(), rs)).FromJust();
+
+        if (rs != nullptr) {
```
Can this be combined into the previous `if (rs != nullptr)` check instead of checking it twice?

sam-github commented:
@jasnell PTAL

Trott commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/19642/

addaleax commented on src/tls_wrap.cc:
```diff
@@ -353,9 +354,41 @@ Local<Value> TLSWrap::GetSSLError(int status, int* err, std::string* msg) {
 
+        Environment* e = env();
+        Isolate* i = e->isolate();
+        Local<Context> c = i->GetCurrentContext();
```
We always call these `env`, `isolate`, and `context`, and I would strongly prefer to keep it this way here; `env` can probably just stay `env()`.

sam-github commented:
@danbev Oops, reordered the error number peek to after the error message long-form printf while addressing some comments, which doesn't work (printf clears the err number). Tests are passing again locally, lets see how CI does.

addaleax commented:
Should this be backported to `v11.x-staging`? If yes please follow [the guide](https://github.com/nodejs/node/blob/master/doc/guides/backporting-to-release-lines.md) and raise a backport PR, if not let me know or add the `dont-land-on` label.

sam-github commented:
Depends on https://github.com/nodejs/node/pull/24729, semver-major, so not a backport candidate.
